src/shared.c

```diff
@@ -696,248 +707,557 @@ bool write_edited_desktop_file(GKeyFile *key_file_structure, const char* appimag
-        /* Some AppImages only have the icon in their root directory, so we have to get it from there */
-        if(desktop_icon_value_original){
-            if((g_str_has_prefix(filename, desktop_icon_value_original)) && (! strstr(filename, "/")) && ( (g_str_has_suffix(filename, ".png")) || (g_str_has_suffix(filename, ".xpm")) || (g_str_has_suffix(filename, ".svg")) || (g_str_has_suffix(filename, ".svgz")))){
-                gchar *file_extension = get_file_extension(filename);
-                gchar *dest_basename = g_strdup_printf("%s_%s_%s.%s", vendorprefix, md5, desktop_icon_value_original, file_extension);
+
+        if (desktop_icon_value_original != NULL) {
+            if (g_str_has_prefix(filename, desktop_icon_value_original) && !strstr(filename, "/") && (g_str_has_suffix(filename, ".png") || g_str_has_suffix(filename, ".xpm") || g_str_has_suffix(filename, ".svg") || g_str_has_suffix(filename, ".svgz"))) {
+                gchar* file_extension = get_file_extension(filename);
+                gchar* dest_basename = g_strdup_printf("%s_%s_%s.%s", vendorprefix, md5, desktop_icon_value_original,
+                    file_extension);
+                g_free(file_extension);
+
+                // clean up dest in case it has been set before
                 g_free(dest);
+
+                // TODO: avoid cluttering /tmp too much
                 dest = g_build_path("/", "/tmp", dest_basename, NULL);
+
                 g_free(dest_basename);
-                g_free(file_extension);
             }
         }
 
+        // cleanup
         g_free(filename);
 
-        if(dest){
-
-            if(verbose)
+        if (dest != NULL) {
+            if (verbose)
                 fprintf(stderr, "install: %s\n", dest);
 
-            gchar *dirname = g_path_get_dirname(dest);
-            if(g_mkdir_with_parents(dirname, 0755))
-                fprintf(stderr, "Could not create directory: %s\n", dirname);
+            gchar* dir_name = g_path_get_dirname(dest);
 
-            g_free(dirname);
+            if (g_mkdir_with_parents(dir_name, 0755)) {
+                fprintf(stderr, "Could not create directory: %s\n", dir_name);
 
-            FILE *f;
-            f = fopen(dest, "w+");
+                // cleanup
+                g_free(dir_name);
 
-            if (f == NULL){
-                fprintf(stderr, "fopen error\n");
+                errored = true;
                 break;
             }
 
-            int r;
-            const void *buff;
+            g_free(dir_name);
+
+            FILE *f = fopen(dest, "w+");
+
+            if (f == NULL) {
+                int error = errno;
+                fprintf(stderr, "Could not open file %s for writing: %s\n", dest, strerror(error));
+
+                // cleanup
+                g_free(dest);
+
+                errored = true;
+                break;
+            }
+
+            const void* buff;
+
             size_t size;
             int64_t offset;
 
-
             for (;;) {
```

Comment:
use "while". Please be explicit.

Reply:
I don't see any implicitness. It's a valid loop, and every software developer should be comfortable with the syntax, and know this is an infinite loop.

I assume `for (;;)` has been used initially because it's (technically) faster than a `while (true)` loop (since in the for loop, there's no checks, whereas in the while version, true would be evaluated in every step -- to be fair, every compiler should be able to optimize out the check in `while (true)`...).